Miner crashing due to unexpected reboot

Due to some hardware issues, my miner machine rebooted while running.  After that, the miner program always ran into crash. The log showed a sector (No. 999) which might not got sealed properly before the reboot and caused the crash.

Is there any code fix to do some checks ahead to avoid crash? 

**The miner log:**

2020-06-10T14:06:13.427+0800    INFO    sectors Proving sector 999
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x50 pc=0xb4f3ea]

goroutine 3819 [running]:
github.com/filecoin-project/storage-fsm.(*Sealing).checkCommit(0xc001ea3200, 0x2b055a0, 0xc00003e090, 0xc006fdbfb0, 0xc, 0x7b4, 0x7, 0xc0096d3720, 0x1, 0x1, ...)
        /home/dianduidian/go/pkg/mod/github.com/filecoin-project/storage-fsm@v0.0.0-20200605082304-aa405b2176aa/checks.go:112 +0x21a
github.com/filecoin-project/storage-fsm.(*Sealing).handleCommitFailed(0xc001ea3200, 0x2b055a0, 0xc00003e090, 0xc00a2ac2b0, 0xc006fdbfb0, 0xc, 0x7b4, 0x7, 0xc0096d3720, 0x1, ...)
        /home/dianduidian/go/pkg/mod/github.com/filecoin-project/storage-fsm@v0.0.0-20200605082304-aa405b2176aa/states_failed.go:162 +0x44b
github.com/filecoin-project/storage-fsm.(*Sealing).Plan.func1(0x2b055a0, 0xc00003e090, 0xc00a2ac2b0, 0xc006fdbfb0, 0xc, 0x7b4, 0x7, 0xc0096d3720, 0x1, 0x1, ...)
        /home/dianduidian/go/pkg/mod/github.com/filecoin-project/storage-fsm@v0.0.0-20200605082304-aa405b2176aa/fsm.go:24 +0x81
reflect.Value.call(0x2445320, 0xc009005c10, 0x13, 0x2718920, 0x4, 0xc00a3d6f70, 0x2, 0x2, 0x51486f, 0x252c760, ...)
        /usr/lib/go-1.14/src/reflect/value.go:460 +0x8ab
reflect.Value.Call(0x2445320, 0xc009005c10, 0x13, 0xc00214a770, 0x2, 0x2, 0x1, 0x0, 0xc000740000)
        /usr/lib/go-1.14/src/reflect/value.go:321 +0xb4
github.com/filecoin-project/go-statemachine.(*StateMachine).run.func3(0xc009005b30, 0x2b055a0, 0xc00003e090, 0xc00a2ac2b0, 0xc009005b40, 0xc006385f20)
        /home/dianduidian/go/pkg/mod/github.com/filecoin-project/go-statemachine@v0.0.0-20200226041606-2074af6d51d9/machine.go:91 +0x2bf
created by github.com/filecoin-project/go-statemachine.(*StateMachine).run
        /home/dianduidian/go/pkg/mod/github.com/filecoin-project/go-statemachine@v0.0.0-20200226041606-2074af6d51d9/machine.go:89 +0x3a0

**The errorneous sector directory:**
![image](https://user-images.githubusercontent.com/4124631/84241693-68039080-ab32-11ea-8b1d-a23dc695df9f.png)

**Other sane sectors:**
![image](https://user-images.githubusercontent.com/4124631/84242031-d6e0e980-ab32-11ea-9395-e49bd6dd32cd.png)

**lotus version:**
lotus version 0.4.0+git.96a5e579.dirty

--- a/checks.go
+++ b/checks.go
@@ -29,6 +29,7 @@ type ErrPrecommitOnChain struct{ error }
 
 type ErrBadSeed struct{ error }
 type ErrInvalidProof struct{ error }
+type ErrNoPrecommit struct{ error }
 
 func checkPieces(ctx context.Context, si SectorInfo, api SealingAPI) error {
 	tok, height, err := api.ChainHead(ctx)
@@ -107,6 +108,10 @@ func (m *Sealing) checkCommit(ctx context.Context, si SectorInfo, proof []byte, 
 	pci, err := m.api.StateSectorPreCommitInfo(ctx, m.maddr, si.SectorNumber, tok)
 	if err != nil {
 		return xerrors.Errorf("getting precommit info: %w", err)
+	}
+
+	if pci == nil {
+		return &ErrNoPrecommit{xerrors.Errorf("precommit info not found on-chain")}
 	}
 
 	if pci.PreCommitEpoch+miner.PreCommitChallengeDelay != si.SeedEpoch {

--- a/fsm.go
+++ b/fsm.go
@@ -96,6 +96,7 @@ var fsmPlanners = map[SectorState]func(events []statemachine.Event, state *Secto
 		on(SectorRetryWaitSeed{}, WaitSeed),
 		on(SectorRetryComputeProof{}, Committing),
 		on(SectorRetryInvalidProof{}, Committing),
+		on(SectorRetryPreCommitWait{}, PreCommitWait),
 	),
 	FinalizeFailed: planOne(
 		on(SectorRetryFinalize{}, FinalizeSector),

--- a/fsm_events.go
+++ b/fsm_events.go
@@ -201,6 +201,10 @@ type SectorRetryWaitSeed struct{}
 
 func (evt SectorRetryWaitSeed) apply(state *SectorInfo) {}
 
+type SectorRetryPreCommitWait struct{}
+
+func (evt SectorRetryPreCommitWait) apply(state *SectorInfo) {}
+
 type SectorRetryComputeProof struct{}
 
 func (evt SectorRetryComputeProof) apply(state *SectorInfo) {

--- a/states.go
+++ b/states.go
@@ -214,9 +214,9 @@ func (m *Sealing) handleWaitSeed(ctx statemachine.Context, sector SectorInfo) er
 		}
 		rand, err := m.api.ChainGetRandomness(ectx, tok, crypto.DomainSeparationTag_InteractiveSealChallengeSeed, randHeight, buf.Bytes())
 		if err != nil {
-			err = xerrors.Errorf("failed to get randomness for computing seal proof: %w", err)
+			err = xerrors.Errorf("failed to get randomness for computing seal proof (ch %d; rh %d; tsk %x): %w", curH, randHeight, tok, err)
 
-			_ = ctx.Send(SectorFatalError{error: err})
+			_ = ctx.Send(SectorChainPreCommitFailed{error: err})
 			return err
 		}
 

--- a/states_failed.go
+++ b/states_failed.go
@@ -177,6 +177,9 @@ func (m *Sealing) handleCommitFailed(ctx statemachine.Context, sector SectorInfo
 			}
 
 			return ctx.Send(SectorRetryInvalidProof{})
+		case *ErrPrecommitOnChain:
+			log.Errorf("no precommit on chain, will retry: %+v", err)
+			return ctx.Send(SectorRetryPreCommitWait{})
 		default:
 			return xerrors.Errorf("checkCommit sanity check error: %w", err)
 		}

